Remotion/ObjectBinding/Web/UI/Controls/Validation/BocListValidationFailureRepository.cs

```diff
@@ -0,0 +1,234 @@
+      }
+
+      public bool HasValidationFailures ()
+      {
+        return _rowFailures is { Count: > 0 } || _cellFailures.Count > 0 && _cellFailures.Values.SelectMany(e => e).Any();
+      }
+
+      public IEnumerable<HandleableFailure> GetAllCellFailuresForRow ()
+      {
+        return _cellFailures.SelectMany(c => c.Value).ToList();
+      }
+    }
+
+    private readonly List<HandleableFailure> _listValidationFailures = new();
+    private readonly Dictionary<IBusinessObject, RowFailureContainer> _rowValidationFailures = new();
+    public BocListValidationFailureRepository ()
+    {
+    }
+
+    public void AddValidationFailuresForBocList (IEnumerable<BusinessObjectValidationFailure> validationFailures)
+    {
+      ArgumentUtility.CheckNotNull(nameof(validationFailures), validationFailures);
+
+      _listValidationFailures.AddRange(validationFailures.Select(e => new HandleableFailure(e)));
+    }
+
+    public void AddValidationFailuresForDataRow (IBusinessObject rowObject, IEnumerable<BusinessObjectValidationFailure> validationFailures)
+    {
+      ArgumentUtility.CheckNotNull(nameof(rowObject), rowObject);
+      ArgumentUtility.CheckNotNull(nameof(validationFailures), validationFailures);
+
+      var failureContainer = GetOrCreateRowFailureContainer(rowObject);
+      failureContainer.AddRowFailures(validationFailures);
+    }
+
+    public void AddValidationFailuresForDataCell (IBusinessObject rowObject, IBocColumnDefinitionWithValidationSupport columnDefinition, IEnumerable<BusinessObjectValidationFailure> validationFailures)
+    {
+      ArgumentUtility.CheckNotNull(nameof(rowObject), rowObject);
+      ArgumentUtility.CheckNotNull(nameof(columnDefinition), columnDefinition);
+      ArgumentUtility.CheckNotNull(nameof(validationFailures), validationFailures);
+
+      var failureContainer = GetOrCreateRowFailureContainer(rowObject);
+      failureContainer.AddCellFailures(columnDefinition, validationFailures);
+    }
+
+
+    public IReadOnlyCollection<BusinessObjectValidationFailure> GetUnhandledValidationFailuresForBocList (bool markAsHandled)
+    {
+      var failures = GetUnhandledValidationFailuresForBocListInternal();
+
+      if (markAsHandled)
+        failures = failures.ApplySideEffect(f => f.MarkAsHandled());
+
+      return failures.Select(f => f.Failure).ToList();
+    }
+
+    public IReadOnlyCollection<BusinessObjectValidationFailure> GetValidationUnhandledFailuresForBocListAndContainingDataRowsAndDataCells (bool markAsHandled)
+    {
+      var failures = GetUnhandledValidationFailuresForBocListInternal();
+
+      foreach (var failureContainer in _rowValidationFailures.Values)
+      {
+        failures = failures.Concat(failureContainer.GetValidationFailuresForDataRow().Where(f => !f.Handled));
+        failures = failures.Concat(failureContainer.GetAllCellFailuresForRow().Where(f => !f.Handled));
+      }
```

Comment:
```suggestion
      var failures = _listValidationFailures
          .Concat(_rowValidationFailures.Values.SelectMany(row => row .GetValidationFailuresForDataRow().Concat(row.GetAllCellFailuresForRow()))
          .Where(f => !f.Handled);
```

Another design-study for your consideration. The exact number of line breaks is something that can be subject to preference. I think this version is the most easiest to read because it has three top level operations and three lines. adding formatting linebreaks in line 2 could also work, it's a matter of preference.